### ISSUE TITLE
Black sky above ground (especially in bottom parts of floatlands)

### ISSUE THREAD

**Issue by Fixer-007** [minetest/minetest#3577]
There is an algorithm that darkens the sky when you go into the caves, but sometimes it fails above and below ground (especially in bottom parts of v7 floatlands). Example:
![alt text](https://i.imgur.com/ADvujnE.png)
P.S. I've removed part about fly sky brightening from this issue. Old topic: https://forum.minetest.net/viewtopic.php?f=6&t=13511

**Comment by paramat** [comment]
Hybrid Dog > Did you disable noclip and fly when you took the screenshot?

Fixer > yes

Doing that lights up the skybox i think.

**Comment by celeron55** [comment]
I'll explain this a bit, for anyone interested.

How the algorithm works is that it detects what the general light level is at where you are looking at along all of the field of view, and makes the background color and fog either bright or dark depending on that.

It's much better than what Minecraft has (or had when I last checked), even when it sometimes can't be perfect.

In the first screenshot there probably is a massive dark cave inside that mountain that the algorithm isn't detecting to be not visible, and in the second screenshot there probably is a thin wall behind that cave with bright sunshine behind.

If you can't stand it, then get a computer with enough performance so that the fog and background color isn't visible. Then this algorithm doesn't have to be used by Minetest.

You can check the code at https://github.com/minetest/minetest/blob/master/src/clientmap.cpp#L661 - as you can see, it's very heuristic-based and full of magic numbers so that it can be as fast as possible. You don't want this code to bring down your FPS because that would be a complete waste of resources.

**Comment by C1ffisme** [comment]
@celeron55 Hmm... That's very interesting. I remember seeing this bug before back in 0.4.2 or 0.4.3, but not knowing what caused it. Thank you for the clarity.

**Comment by Fixer-007** [comment]
>and in the second screenshot there probably is a thin wall behind that cave with bright sunshine behind.

second screenshot was deep underground, disabling the fog will darken it fully.

**Comment by paramat** [comment]
Can we close this? Or keep open if you have an improvement to the code in mind and wish to discuss.

**Comment by paramat** [comment]
Closing, see celeron55's explanation.

**Comment by Fixer-007** [comment]
Another nice example of the problem: https://i.imgur.com/GLQcc0l.png

**Comment by paramat** [comment]
@Fixer-007 you might have freemove enabled but without fly privilege, this happened to me too in the nether.

**Comment by Fixer-007** [comment]
You are right, thats embaracing and also crazy

**Comment by Fixer-007** [comment]
About first pic - this black background thing appears a lot when exploring new v7 floatlands (especially below).

**Comment by pilino1234** [comment]
This happens a lot on latest dev, even when moving around (no caves, floatlands, etc.). Added some screenshots below showing how the slightest mouse movement can trigger this. This isn't really [Low priority], more like [Blocker] since it really affects the experience. "Why does the sky turn black if I point at this node?!"

Pic1: Normal (clouds visible, sky is blue)
![screenshot_20161202_215517](https://cloud.githubusercontent.com/assets/11665015/20861070/31ce7d90-b987-11e6-8554-948a9f407923.png)

Pic2: Clouds are gone, sky is grey
![screenshot_20161202_215524](https://cloud.githubusercontent.com/assets/11665015/20861072/43039a96-b987-11e6-9096-2f27f8910ffe.png)

----
Pic3: Looks normal (blue sky with clouds)
![screenshot_20161202_215534](https://cloud.githubusercontent.com/assets/11665015/20861076/51353296-b987-11e6-8184-8d20a173b891.png)

Pic4: Sky turns black, despite time of day and lighting for the rendered world
![screenshot_20161202_215540](https://cloud.githubusercontent.com/assets/11665015/20861078/5e6eb20c-b987-11e6-914d-4b114f6ae81c.png)

----
Pic5&6: sky is normal (5: normal rendering, 6: wireframe (there are no caves behind the pointer))
![screenshot_20161203_181350](https://cloud.githubusercontent.com/assets/11665015/20861092/b6614844-b987-11e6-8c26-a2cddcf56026.png)
![screenshot_20161203_181357](https://cloud.githubusercontent.com/assets/11665015/20861093/b7cd90b6-b987-11e6-8411-4b6187afc08b.png)

Pic7&8: sky is grey, no clouds shown (7: normal rendering, 8: wireframe)
![screenshot_20161203_181413](https://cloud.githubusercontent.com/assets/11665015/20861095/bc9e00ee-b987-11e6-8dd2-21c5e737d712.png)
![screenshot_20161203_181415](https://cloud.githubusercontent.com/assets/11665015/20861096/be0a9e10-b987-11e6-8c1e-b07f48808473.png)

Pic9&10: sky is black, no clouds shown (9: normal rendering, 10: wireframe)
![screenshot_20161203_181447](https://cloud.githubusercontent.com/assets/11665015/20861097/c3c4720e-b987-11e6-8cdb-53d5e67eea70.png)
![screenshot_20161203_181450](https://cloud.githubusercontent.com/assets/11665015/20861098/c4e09780-b987-11e6-9178-f006d07aa964.png)

**Comment by paramat** [comment]
Sorry i closed this.
I have noticed this is a problem below mgv7 floatlands, because it is a large volume of complete darkness the black sky is triggererd, correctly, but looks wrong.

**Comment by paramat** [comment]
PR for underneath floatlands #5028

**Comment by Xanthin** [comment]
Still present for me since around #4646, even with `nocaves,nofloatlands,nocaverns`.

**Comment by paramat** [comment]
Yes it will be, #5028 was not merged as the bright underside it caused does not look good. So not sure what to do about this.

**Comment by Fixer-007** [comment]
There is another problem with this algorithm (quoting paramat):
>clouds disappear whenever 'darkness detection' decides that sky is not visible. If you could see the sky it would be going dark at these times.

**Comment by paramat** [comment]
^ But that's correct behaviour.

**Comment by Fixer-007** [comment]
But false-positives above make it a problem.

**Comment by paramat** [comment]
Yes, but it's not an additional problem to dark sky, the same variable that makes clouds disappear makes sky go dark. Dark sky is already bad enough so clouds disappearing at the same time is no additional issue.

**Comment by Fixer-007** [comment]
Situation was improved, if it gets annoying, i will reopen.

**Comment by paramat** [comment]
#6117 merged.

**Comment by Fixer-007** [comment]
And I forgot to check floatlands... will recheck later.

**Comment by paramat** [comment]
Floatlands will still have a problem because there is large volumes of dark air beneath them, but htat isn't a bug with the darkness detection code, the game inevitably thinks it is in a big underground cave.

**Comment by Fixer-007** [comment]
Still occurs under floatlands :( closed too early.

**Comment by paramat** [comment]
I do know of a fix for floatlands but it's a bit ugly: disabling shadow propagation from floatland midpoint down to lowest point.

**Comment by Zughy** [comment]
https://github.com/minetest/minetest/issues/14468#issuecomment-2298703749

**Comment by appgurueu** [comment]
I don't think this issue has since been fixed, though it may be less severe these days.

Fixing the aforementioned algorithm (https://github.com/minetest/minetest/issues/3577#issuecomment-196048898) is probably nontrivial (esp. considering the breadth of possible configurations), but we could at least let mods disable this behavior via a sky parameter. Mods could still darken the sky themselves if they want to, implementing more suitable heuristics.

**Comment by nininik0** [comment]
This lowkey is still not fixed, like DAYUM this issue is 8 YEARS NOW-

**Comment by nininik0** [comment]
How is this still not fixed 😭- happens on 5.15.1 still as of 2026.

**Comment by CrazyladMT** [comment]
It was cool to just now read the entire issue and see it get closed and then re-opened, and then closed again and re-opened again. 

I agree with @appgurueu's proposed solution—letting modders disable this via a sky parameter seems very helpful.

**Comment by PookaMustard** [comment]
This issue has not been fixed, it occurs for me regularly in the latest Mineclonia, with both the latest stable Luanti engine on Windows and Linux, and the development branch on Linux. I think it also occurs in Android with the F-Droid build but I don't recall when the last time it happened. It might be less severe than in the original 2016 issue because I can only see the black sky at angles further downward than the OP.

**Comment by appgurueu** [comment]
You could ask the Mineclonia developers whether they find #16976 useful. It allows turning off the engine algorithm entirely, game developers can then implement whatever they find suitable instead. In the future, SSCSM will also let them implement clientside logic.

The bugs in this algorithm have indeed not been fixed, and potentially never will be (this is tagged low-priority and non-trivial after all), which is why this issue is still open.
